akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -136,3 +139,59 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueue[T]) extend
   def offer(elem: T): CompletionStage[QueueOfferResult] = delegate.offer(elem).toJava
   def watchCompletion(): CompletionStage[Done] = delegate.watchCompletion().toJava
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class BlockingStreamSource[T, S](create: () ⇒ S,
```

Comment:
> For the asynchronous variant we could probably get away with a function S => Future[T] and a object EOF extends RuntimeException with NoStackTrace.

@rkuhn, We have headOption and lastOption as Future[Option[T]]. Should we stick to this? I don't like idea of throwing exception when the case is not exceptional by nature.